api/layer/object.go

```diff
@@ -363,22 +353,47 @@ func (n *layer) listSortedAllObjects(ctx context.Context, p allObjectParams) ([]
 	return objects, nil
 }
 
-func trimStartAfter(startAfter string, objects []*ObjectInfo) []*ObjectInfo {
-	if objects != nil && len(startAfter) != 0 && objects[0].Name <= startAfter {
-		for i := range objects {
-			if objects[i].Name > startAfter {
-				return objects[i:]
-			}
+func trimAfterObjectName(startAfter string, objects []*ObjectInfo) []*ObjectInfo {
+	if len(objects) != 0 && objects[len(objects)-1].Name <= startAfter {
+		return nil
+	}
+	for i := range objects {
+		if objects[i].Name > startAfter {
+			return objects[i:]
 		}
 	}
-	return objects
+
+	return nil
 }
 
 func trimAfterObjectID(id string, objects []*ObjectInfo) []*ObjectInfo {
+	if len(objects) != 0 && objects[len(objects)-1].id.String() == id {
+		return nil
+	}
 	for i, obj := range objects {
 		if obj.ID().String() == id {
-			return objects[i:]
+			return objects[i+1:]
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) trimAndCacheRestObjects(cacheKey string, maxKeys int, objects []*ObjectInfo) []*ObjectInfo {
+	restObjects := objects[maxKeys-1:]
+	n.cache.Put(cacheKey, restObjects)
```

Comment:
I am not sure about cashing trimmed result. Consider bucket `XXX` with 100 keys. There are two consecutive requests from the same user:

1. `ListObjectV1` of bucket `XXX` with `MaxKeys` limit 10
2. `ListObjectV2` of bucket `XXX` with `MaxKeys` limit 100

First request will cache remaining 90 keys after the trim.
Second request will return 90 keys from the cache, instead of 100, which is incorrect.

Am I missing something?

Reply:
I thought we search for a cache entry when a user sets ContinuationToken/Marker in a request, so the user should get a response from cache if he received the token from the previous response. That's why cache doesn't need to keep objects before ContinuationToken/Marker (it makes search by the token a little faster).
 
But I didn't thought about different versions compatibility. I have some suspicions this implementation will not work in this case correctly,  because, yes, you're right,  ListObjectV1 will put a "broken" (i mean filtered with old markers)  cache, which is useless for another version. 

There is two possible ways: 
1. Somehow separate entries in a map of cache for different versions
2. Create a common cache with unfiltered objects for all versions, get back workers and remove afterfuncs, because with afterfunc the cache entry lives only for 1 minute after creating of the entry and we can't prolong existence, but it'd be better to add a way to update timer for cache entry, if we want to read this cache from different ListObjects versions.

Reply:
> 1. Somehow separate entries in a map of cache for different versions

In worst case scenario we will have two different caches with the same data, sounds bad.

> 2. Create a common cache with unfiltered objects for all versions,

I would go for it.

>  get back workers and remove afterfuncs, because with afterfunc the cache entry lives only for 1 minute after creating of the entry and we can't prolong existence, but it'd be better to add a way to update timer for cache entry, if we want to read this cache from different ListObjects versions.

Cache invalidation is one of the hardest problems in CS. With proposed prolongation there is a chance we will never invalidate the cache at all. I propose to keep things as they are right now and go back to the cache prolongation issues when we will be tuning performance of the S3 gate.